src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs

```diff
@@ -1639,13 +1652,23 @@ private async Task ClearMarkers(ScriptFile scriptFile, EventContext eventContext
                 });
         }
 
+        private static string GetUniqueIdFromDiagnostic(Diagnostic diagnostic)
+        {
+            string source = diagnostic.Source ?? "?";
+            string code = diagnostic.Code ?? "?";
+            string severity = diagnostic.Severity != null ? diagnostic.Severity.ToString() : "?";
+            Position start = diagnostic.Range.Start;
+            Position end = diagnostic.Range.End;
+            return $"{source}_{code}_{severity}_{start.Line}:{start.Character}-{end.Line}:{end.Character}";
```

Comment:
I have a mild trepidation about allocating a string ID for every diagnostic marker, but I think it's definitely worth a shot -- we'll soon find out if the cache is worth it.

A couple of alternatives might be:
1. Use a string builder to create the ID (I think there are documented advantages over a string template, but theoretically it seems like a string template should be just as performant).
2. Combine hashcodes of the elements. Only problem being that .NET 451 and .NET Standard 2.0 don't have a nice `HashCode.Combine()` method like .NET Core 2.1 does.

Reply:
Hmm, looking at some internet perf reviews of string interpolation in C# ... OK will switch to StringBuilder.   :-) 

BTW the previous approach allocated a string as well - `Guid.NewGuid().ToString()`.